### ISSUE TITLE
Control of random generator

### ISSUE THREAD

**Issue by lionelkusch** [mind-inria/hidimstat#360]
For all the children classes of `BasePerturbation`, the randomness only appears at the `.predict()` step. So, for clarity, I suggest we only set/check `random_state` starting from that stage. 

In addition, it would be desirable that the behavior of the fitted class is deterministic: calling twice  `.importance()` on the same data gives the same result. 

I suggest that we expose the `random_state` at the importance/predict methods, but not before. 

For compatibility with CFI, this requires modifying the `ConditionalSampler`, which currently takes the `random_state` at __init__ but only uses it at `.sample()` we could thus make it an argument of `.sample(random_state=None)`

_Originally posted by @jpaillard in https://github.com/mind-inria/hidimstat/pull/344#discussion_r2307234877_

**Comment by lionelkusch** [comment]
Agreed. However, exposing the random_state and setting its default value to None keeps this as a default behavior. But it allows, in addition, to control the randomness if desired. 

From the developer's perspective, I think it would clarify the code, narrowing the sections where randomness appears/is controlled.

_Originally posted by @jpaillard in https://github.com/mind-inria/hidimstat/pull/344#discussion_r2307282129_

**Comment by jpaillard** [comment]
Sorry if I was not clear but [`GaussianMixture` does what I try to describe](https://github.com/scikit-learn/scikit-learn/blob/c5497b7f7eacfaff061cf68e09bcd48aa93d4d6b/sklearn/mixture/_base.py#L437).

```python
class BaseMixture() ...
   ...
   def __init__(self, random_state=None, ...):
      self.random_state = random_state
      ...

   def sample(self, ...): 
      rng = check_random_state(self.random_state) 
...
```

In this way, if you let `random_state=None` multiple calls of sample lead to different random draws. But if you pass an int you get a deterministic behavior. 
You can try this snippet to check: 

``` python 
import numpy as np
from sklearn.mixture import GaussianMixture

for random_state in [0, None]:
    X = np.random.rand(100, 2)
    mixture = GaussianMixture(n_components=2, random_state=random_state)
    mixture.fit(X)

    samples_1 = mixture.sample(n_samples=10)
    samples_2 = mixture.sample(n_samples=10)

    eq_samples = np.array_equal(samples_1[0], samples_2[0])
    print(f"Random state: {random_state}, Equal samples: {eq_samples}")
```

It is then just a question of whether we want to expose the random_state in the `__init__` or `.predict()`, I am OK with both, 


## Concerning D0CRT
the random_state parameter that is passed in the __init__ is actuallt never used. The only random_state that are passed are those of the sklearn linear models. For me this is OK, it is just to control the splitters, otherwise the estimators behave in the way I described.

**Comment by lionelkusch** [comment]
My underling question is: How to control the random state of the internal model?
As example: LassoCV for DCRT or desparsifiedLasso.

By looking at the code of LassoCV or Lasso, it require the definition of the random generator during the fit method (see line:https://github.com/scikit-learn/scikit-learn/blob/c5497b7f7eacfaff061cf68e09bcd48aa93d4d6b/sklearn/linear_model/_coordinate_descent.py#L636).

Here, the difficulty is that the model are fitted in parallel.
My actual solution that I implemented in CFI (in PR #344) is to generate a sequence of number base on 1 number generated randomly.

**Comment by bthirion** [comment]
Note that you want computed importances to be made deterministic for a debugging prupose: if you onbtain a counter-intyitive result, you want to be able to reproduce it to introspect what's going on.

**Comment by lionelkusch** [comment]
In the two cases, the behaviour is totally reproducible and controllable. The only difference is the expectation of the users for the method `importance` in the case of CFI or PFI which is based on randomness.
The second point is the homogenisation of the control of the random generator, which can be used in `fit` in some cases.

**Comment by lionelkusch** [comment]
I want to have control of randomness. This is why I need to seed a seed but I also want to have different results when I run multiple times the same function to identify, for example, a confidence interval of the result.
With the solution proposed by @jpaillard, I need to pass a random generator but the reset of the random generator is, for me, a bit hidden by the method. In my case, I let the user do the reset of the generator when it wants, except at the beginning.

**Comment by jpaillard** [comment]
As said in the PR, I suggest we follow the implementation from sklearn's RandomForest, which presents a similar seeding question. The logic is presented in [`_make_estimator`](https://github.com/scikit-learn/scikit-learn/blob/91d5640a82dd0bd35c8c438ef22ff3c2cd56bce3/sklearn/ensemble/_base.py#L141) in the sub-function `_set_random_states`. It uses a seed based on a random generator. 

> Moreover, the usage of a seed based on a random generator is a problem because the chance of having 2 times the same seed is highly probable. This problem is called the [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem).


For me, this is not a problem.  Seeds can be sampled uniformly with replacement. It's fine to end up with 2 `ConditionalSampler` that have the same seed.

**Comment by bthirion** [comment]
> For all the children classes of `BasePerturbation`, the randomness only appears at the `.predict()` step. So, for clarity, I suggest we only set/check `random_state` starting from that stage.
> 
> In addition, it would be desirable that the behavior of the fitted class is deterministic: calling twice `.importance()` on the same data gives the same result.
> 
> I suggest that we expose the `random_state` at the importance/predict methods, but not before.
> 
> For compatibility with CFI, this requires modifying the `ConditionalSampler`, which currently takes the `random_state` at **init** but only uses it at `.sample()` we could thus make it an argument of `.sample(random_state=None)`
> 
> _Originally posted by [@jpaillard](https://github.com/jpaillard) in [#344 (comment)](https://github.com/mind-inria/hidimstat/pull/344#discussion_r2307234877)_

I'm afraid I don't understand you suggestion.
the `random_state` or `random_generator`, has to be given at class instantiation. Then it's OK if it is used only when  `predict` is called. What is the problem with this pattern ?

**Comment by jpaillard** [comment]
The issue we have is that, currently, at instantiation, `self.radom_state = check_ranom_state(rando_state)` creates a Generator object. Then, each call of 'predict' modifies its state, which prevents getting deterministic results.

The suggestion is: 
At instantiation, `self.radom_state=self.radom_state` 

Then, in predict 
`rng=check_random_state(self.radom_state)`  and use the rng for permutations ... 

That way, users can obtain either deterministic or randomness in their results depending on the type passed for random_state

**Comment by bthirion** [comment]
OK, that makes sense. Thank you for the explanations.

**Comment by lionelkusch** [comment]
> As said in the PR, I suggest we follow the implementation from sklearn's RandomForest, which presents a similar seeding question. The logic is presented in [`_make_estimator`](https://github.com/scikit-learn/scikit-learn/blob/91d5640a82dd0bd35c8c438ef22ff3c2cd56bce3/sklearn/ensemble/_base.py#L141) in the sub-function `_set_random_states`. It uses a seed based on a random generator.
> 
> > Moreover, the usage of a seed based on a random generator is a problem because the chance of having 2 times the same seed is highly probable. This problem is called the [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem).
> 
> For me, this is not a problem. Seeds can be sampled uniformly with replacement. It's fine to end up with 2 `ConditionalSampler` that have the same seed.

The code for setting the seed is 9 years old and they didn't update it with the new functionalities of Numpy. Moreover, I think nobody looks that 2 trees can have the same seed.

For me, this is problematic and there are solutions to solving it. I don't see why you don't want it.
